storage/innobase/dict/dict0stats.cc

```diff
@@ -1372,6 +1379,77 @@ static dberr_t btr_pcur_open_level(btr_pcur_t *pcur, ulint level, mtr_t *mtr,
 }
 
 
+/** Stores various statistical metrics gathered during an index scan
+or fetch operation. */
+struct IndexScanStatistics
+{
+  /* Array for number of distinct keys for all prefixes */
+  ib_uint64_t *n_diff_on_level = nullptr;
+
+  /* Total number of records */
+  ib_uint64_t total_recs = 0;
+
+  /* Total number of pages */
+  ib_uint64_t total_pages = 0;
+
+  /* Boundaries of the groups of distinct keys */
+  boundaries_t *n_diff_boundaries = nullptr;
+
+  /* Array of number of non-null keys for all prefixes */
+  ib_uint64_t *n_non_null_on_level = nullptr;
+
+  /* Whether nulls are considered unequal for statistics */
+  bool nulls_unequal = false;
+
+  IndexScanStatistics(bool nulls_unequal_): nulls_unequal(nulls_unequal_) {}
+  IndexScanStatistics(bool nulls_unequal_, bool non_null,
+                      dict_index_t *index)
+  : nulls_unequal(nulls_unequal_)
+  {
+    memset(index->stat_n_diff_key_vals, 0x0,
+	   index->n_uniq * sizeof(index->stat_n_diff_key_vals[0]));
+    n_diff_on_level= index->stat_n_diff_key_vals;
+    if (non_null)
+    {
+      memset(index->stat_n_non_null_key_vals, 0x0,
+             index->n_uniq * sizeof(index->stat_n_non_null_key_vals[0]));
+      n_non_null_on_level= index->stat_n_non_null_key_vals;
+    }
+  }
+
+  void init(bool non_null, ulint n_uniq)
+  {
+    if (non_null)
+      n_non_null_on_level= static_cast<ib_uint64_t*>(
+        ut_zalloc(n_uniq * sizeof(ib_uint64_t),
+                  mem_key_dict_stats_n_diff_on_level));
```

Comment:
I find it very confusing that sometimes, the constructor makes `n_non_null_on_level` and `n_diff_on_level` point directly to `dict_index_t`, but after the `init()` call they may point to separate, heap-allocated memory. Especially the call to `destroy()` would seem to able to corrupt things if the pointers were pointing to members of `dict_index_t`:

Why is `init()` not returning any error status? Can we just invoke `calloc()` here and do proper error handling?

Reply:
This had been marked as resolved, but without any comment regarding what was done. I see that the `IndexLevelStats` constructors are invoking `ut_zalloc` in the most recent version.  Maybe this code was moved there.